File: src/services/notification_service.ts
/**
 * Frontend notification service for handling real-time updates.
 * Integrates with WebSocket connection and provides notification management.
 */

import { toast } from 'react-hot-toast';

export interface NotificationData {
  id?: string;
  type: string;
  title: string;
  message: string;
  severity: 'info' | 'success' | 'warning' | 'error';
  duration?: number;
  persistent?: boolean;
  data?: Record<string, any>;
}

export interface WebSocketNotification {
  type: string;
  data: Record<string, any>;
  timestamp: string;
  message_id: string;
}

class NotificationService {
  private notifications: NotificationData[] = [];
  private listeners: ((notification: NotificationData) => void)[] = [];
  private maxNotifications = 100;

  /**
   * Add a new notification
   */
  addNotification(notification: NotificationData): void {
    const notificationWithId = {
      ...notification,
      id: notification.id || this.generateId(),
    };

    this.notifications.unshift(notificationWithId);

    // Limit the number of stored notifications
    if (this.notifications.length > this.maxNotifications) {
      this.notifications = this.notifications.slice(0, this.maxNotifications);
    }

    // Show toast notification
    this.showToast(notificationWithId);

    // Notify listeners
    this.listeners.forEach(listener => listener(notificationWithId));
  }

  /**
   * Process WebSocket notification
   */
  processWebSocketNotification(wsNotification: WebSocketNotification): void {
    const notification = this.mapWebSocketToNotification(wsNotification);
    if (notification) {
      this.addNotification(notification);
    }
  }

  /**
   * Map WebSocket notification to internal notification format
   */
  private mapWebSocketToNotification(wsNotification: WebSocketNotification): NotificationData | null {
    const { type, data } = wsNotification;

    switch (type) {
      case 'agent_run_started':
        return {
          type: 'agent_run',
          title: 'Agent Run Started',
          message: `Started processing: ${data.target_text?.substring(0, 50)}...`,
          severity: 'info',
          data: data,
        };

      case 'agent_run_progress':
        // Don't show toast for progress updates to avoid spam
        return {
          type: 'agent_run_progress',
          title: 'Agent Run Progress',
          message: `Progress: ${Math.round(data.progress)}%${data.current_step ? ` - ${data.current_step}` : ''}`,
          severity: 'info',
          duration: 0, // No toast
          data: data,
        };

      case 'agent_run_completed':
        return {
          type: 'agent_run',
          title: 'Agent Run Completed',
          message: 'Agent run completed successfully',
          severity: 'success',
          data: data,
        };

      case 'agent_run_failed':
        return {
          type: 'agent_run',
          title: 'Agent Run Failed',
          message: data.error || 'Agent run failed',
          severity: 'error',
          persistent: true,
          data: data,
        };

      case 'validation_started':
        return {
          type: 'validation',
          title: 'Validation Started',
          message: `Started validation: ${data.name}`,
          severity: 'info',
          data: data,
        };

      case 'validation_completed':
        return {
          type: 'validation',
          title: 'Validation Completed',
          message: `Validation ${data.success ? 'passed' : 'failed'}: ${data.name || 'Unknown'}`,
          severity: data.success ? 'success' : 'error',
          data: data,
        };

      case 'validation_failed':
        return {
          type: 'validation',
          title: 'Validation Failed',
          message: data.error || 'Validation failed',
          severity: 'error',
          persistent: true,
          data: data,
        };

      case 'project_updated':
        return {
          type: 'project',
          title: 'Project Updated',
          message: 'Project settings have been updated',
          severity: 'info',
          data: data,
        };

      case 'system_alert':
        return {
          type: 'system',
          title: data.alert_type || 'System Alert',
          message: data.message,
          severity: data.severity || 'info',
          persistent: data.severity === 'error',
          data: data,
        };

      case 'user_message':
        return {
          type: 'user_message',
          title: 'Message',
          message: data.message,
          severity: data.message_type || 'info',
          data: data.data,
        };

      case 'connection_established':
        return {
          type: 'connection',
          title: 'Connected',
          message: 'Real-time connection established',
          severity: 'success',
          duration: 2000,
          data: data,
        };

      default:
        console.log('Unhandled WebSocket notification type:', type);
        return null;
    }
  }

  /**
   * Show toast notification
   */
  private showToast(notification: NotificationData): void {
    if (notification.duration === 0) {
      return; // Skip toast
    }

    const options = {
      duration: notification.duration || (notification.persistent ? Infinity : 4000),
      position: 'top-right' as const,
    };

    switch (notification.severity) {
      case 'success':
        toast.success(notification.message, options);
        break;
      case 'error':
        toast.error(notification.message, options);
        break;
      case 'warning':
        toast.error(notification.message, options); // Use error style for warnings
        break;
      case 'info':
      default:
        toast(notification.message, options);
        break;
    }
  }

  /**
   * Get all notifications
   */
  getNotifications(): NotificationData[] {
    return [...this.notifications];
  }

  /**
   * Get notifications by type
   */
  getNotificationsByType(type: string): NotificationData[] {
    return this.notifications.filter(n => n.type === type);
  }

  /**
   * Clear all notifications
   */
  clearNotifications(): void {
    this.notifications = [];
    this.listeners.forEach(listener => listener({ 
      type: 'clear', 
      title: '', 
      message: '', 
      severity: 'info' 
    }));
  }

  /**
   * Clear notifications by type
   */
  clearNotificationsByType(type: string): void {
    this.notifications = this.notifications.filter(n => n.type !== type);
  }

  /**
   * Remove specific notification
   */
  removeNotification(id: string): void {
    this.notifications = this.notifications.filter(n => n.id !== id);
  }

  /**
   * Add listener for notification updates
   */
  addListener(listener: (notification: NotificationData) => void): () => void {
    this.listeners.push(listener);
    
    // Return unsubscribe function
    return () => {
      const index = this.listeners.indexOf(listener);
      if (index > -1) {
        this.listeners.splice(index, 1);
      }
    };
  }

  /**
   * Generate unique ID
   */
  private generateId(): string {
    return `notification_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Show success notification
   */
  success(title: string, message: string, data?: Record<string, any>): void {
    this.addNotification({
      type: 'manual',
      title,
      message,
      severity: 'success',
      data,
    });
  }

  /**
   * Show error notification
   */
  error(title: string, message: string, data?: Record<string, any>): void {
    this.addNotification({
      type: 'manual',
      title,
      message,
      severity: 'error',
      persistent: true,
      data,
    });
  }

  /**
   * Show warning notification
   */
  warning(title: string, message: string, data?: Record<string, any>): void {
    this.addNotification({
      type: 'manual',
      title,
      message,
      severity: 'warning',
      data,
    });
  }

  /**
   * Show info notification
   */
  info(title: string, message: string, data?: Record<string, any>): void {
    this.addNotification({
      type: 'manual',
      title,
      message,
      severity: 'info',
      data,
    });
  }
}

// Export singleton instance
export const notificationService = new NotificationService();
export default notificationService;

File: backend/app/websocket/notification_service.py
"""
Notification service for real-time updates.
Handles different types of notifications and message formatting.
"""

import logging
from typing import Dict, Any, Optional, List
from enum import Enum

from .connection_manager import ConnectionManager, WebSocketMessage

logger = logging.getLogger(__name__)


class NotificationType(Enum):
    """Types of notifications"""
    AGENT_RUN_STARTED = "agent_run_started"
    AGENT_RUN_PROGRESS = "agent_run_progress"
    AGENT_RUN_COMPLETED = "agent_run_completed"
    AGENT_RUN_FAILED = "agent_run_failed"
    VALIDATION_STARTED = "validation_started"
    VALIDATION_COMPLETED = "validation_completed"
    VALIDATION_FAILED = "validation_failed"
    PROJECT_UPDATED = "project_updated"
    SYSTEM_ALERT = "system_alert"
    USER_MESSAGE = "user_message"


class NotificationService:
    """Service for handling real-time notifications"""
    
    def __init__(self, connection_manager: ConnectionManager):
        self.connection_manager = connection_manager
    
    async def notify_agent_run_started(self, 
                                     project_id: str, 
                                     run_id: str, 
                                     target_text: str,
                                     user_id: Optional[str] = None):
        """Notify that an agent run has started"""
        message = WebSocketMessage(
            type=NotificationType.AGENT_RUN_STARTED.value,
            data={
                "run_id": run_id,
                "project_id": project_id,
                "target_text": target_text,
                "status": "started"
            }
        )
        
        await self.connection_manager.broadcast_to_project(project_id, message)
        
        if user_id:
            await self.connection_manager.send_to_user(user_id, message)
        
        logger.info(f"Notified agent run started: {run_id}")
    
    async def notify_agent_run_progress(self, 
                                      project_id: str, 
                                      run_id: str, 
                                      progress: float,
                                      current_step: Optional[str] = None,
                                      user_id: Optional[str] = None):
        """Notify agent run progress update"""
        message = WebSocketMessage(
            type=NotificationType.AGENT_RUN_PROGRESS.value,
            data={
                "run_id": run_id,
                "project_id": project_id,
                "progress": progress,
                "current_step": current_step,
                "status": "running"
            }
        )
        
        await self.connection_manager.broadcast_to_project(project_id, message)
        
        if user_id:
            await self.connection_manager.send_to_user(user_id, message)
    
    async def notify_agent_run_completed(self, 
                                       project_id: str, 
                                       run_id: str, 
                                       result: Dict[str, Any],
                                       user_id: Optional[str] = None):
        """Notify that an agent run has completed"""
        message = WebSocketMessage(
            type=NotificationType.AGENT_RUN_COMPLETED.value,
            data={
                "run_id": run_id,
                "project_id": project_id,
                "result": result,
                "status": "completed"
            }
        )
        
        await self.connection_manager.broadcast_to_project(project_id, message)
        
        if user_id:
            await self.connection_manager.send_to_user(user_id, message)
        
        logger.info(f"Notified agent run completed: {run_id}")
    
    async def notify_agent_run_failed(self, 
                                    project_id: str, 
                                    run_id: str, 
                                    error: str,
                                    user_id: Optional[str] = None):
        """Notify that an agent run has failed"""
        message = WebSocketMessage(
            type=NotificationType.AGENT_RUN_FAILED.value,
            data={
                "run_id": run_id,
                "project_id": project_id,
                "error": error,
                "status": "failed"
            }
        )
        
        await self.connection_manager.broadcast_to_project(project_id, message)
        
        if user_id:
            await self.connection_manager.send_to_user(user_id, message)
        
        logger.error(f"Notified agent run failed: {run_id} - {error}")
    
    async def notify_validation_started(self, 
                                      project_id: str, 
                                      validation_id: str, 
                                      name: str,
                                      agent_run_id: Optional[str] = None):
        """Notify that validation has started"""
        message = WebSocketMessage(
            type=NotificationType.VALIDATION_STARTED.value,
            data={
                "validation_id": validation_id,
                "project_id": project_id,
                "agent_run_id": agent_run_id,
                "name": name,
                "status": "started"
            }
        )
        
        await self.connection_manager.broadcast_to_project(project_id, message)
        
        logger.info(f"Notified validation started: {validation_id}")
    
    async def notify_validation_completed(self, 
                                        project_id: str, 
                                        validation_id: str, 
                                        results: Dict[str, Any],
                                        success: bool = True):
        """Notify that validation has completed"""
        message = WebSocketMessage(
            type=NotificationType.VALIDATION_COMPLETED.value,
            data={
                "validation_id": validation_id,
                "project_id": project_id,
                "results": results,
                "success": success,
                "status": "completed"
            }
        )
        
        await self.connection_manager.broadcast_to_project(project_id, message)
        
        logger.info(f"Notified validation completed: {validation_id} (success: {success})")
    
    async def notify_validation_failed(self, 
                                     project_id: str, 
                                     validation_id: str, 
                                     error: str):
        """Notify that validation has failed"""
        message = WebSocketMessage(
            type=NotificationType.VALIDATION_FAILED.value,
            data={
                "validation_id": validation_id,
                "project_id": project_id,
                "error": error,
                "status": "failed"
            }
        )
        
        await self.connection_manager.broadcast_to_project(project_id, message)
        
        logger.error(f"Notified validation failed: {validation_id} - {error}")
    
    async def notify_project_updated(self, 
                                   project_id: str, 
                                   changes: Dict[str, Any],
                                   user_id: Optional[str] = None):
        """Notify that a project has been updated"""
        message = WebSocketMessage(
            type=NotificationType.PROJECT_UPDATED.value,
            data={
                "project_id": project_id,
                "changes": changes,
                "updated_by": user_id
            }
        )
        
        await self.connection_manager.broadcast_to_project(project_id, message)
        
        logger.info(f"Notified project updated: {project_id}")
    
    async def notify_system_alert(self, 
                                alert_type: str, 
                                message: str, 
                                severity: str = "info",
                                project_id: Optional[str] = None):
        """Send system-wide alert"""
        alert_message = WebSocketMessage(
            type=NotificationType.SYSTEM_ALERT.value,
            data={
                "alert_type": alert_type,
                "message": message,
                "severity": severity,
                "project_id": project_id
            }
        )
        
        if project_id:
            await self.connection_manager.broadcast_to_project(project_id, alert_message)
        else:
            await self.connection_manager.broadcast_to_all(alert_message)
        
        logger.warning(f"System alert ({severity}): {message}")
    
    async def send_user_message(self, 
                              user_id: str, 
                              message: str, 
                              message_type: str = "info",
                              data: Optional[Dict[str, Any]] = None):
        """Send direct message to a specific user"""
        user_message = WebSocketMessage(
            type=NotificationType.USER_MESSAGE.value,
            data={
                "message": message,
                "message_type": message_type,
                "data": data or {}
            }
        )
        
        await self.connection_manager.send_to_user(user_id, user_message)
        
        logger.info(f"Sent user message to {user_id}: {message}")
    
    async def notify_batch_updates(self, 
                                 project_id: str, 
                                 updates: List[Dict[str, Any]]):
        """Send batch updates for efficiency"""
        message = WebSocketMessage(
            type="batch_updates",
            data={
                "project_id": project_id,
                "updates": updates,
                "count": len(updates)
            }
        )
        
        await self.connection_manager.broadcast_to_project(project_id, message)
        
        logger.info(f"Sent batch updates to project {project_id}: {len(updates)} updates")
    
    def get_notification_stats(self) -> Dict[str, Any]:
        """Get notification service statistics"""
        connection_stats = self.connection_manager.get_connection_stats()
        
        return {
            "connection_manager": connection_stats,
            "service_status": "active"
        }
